Feature/trigger channel preference query from bml

morantr opened the pull request:
Add new command to bml and its cli, which triggers channel selection flow on specific son_slave, by sending channel preference query request.

Issue #74

morantr committed:
common: tlvf: add bml message for channel selection

Add a message to yaml files which will be used to trigger
channel selection from bml.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: tlvf: Add auto generated code
Add auto generated code, created by adding  a new message in previous
commit, which triggers channel selection from bml, in previous commit.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: bml: Add Function to bml_internal
Add function to bml_internal,which sends channel selection
request to the controller.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add channel selection function to bml
Add function to bml which calls to channel selection function on
bml_internal.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add channel selection function to CLI
Add function to beerocks_cli that calls bml api, which sends channel
selection request to the controller.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add channel selection command to CLI
Add channel selection command to beerocks_cli_bml command map
and its caller.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Moved controller db function to public
Moved function that create key from ruid and al_mac on controller db
to public instead of private.
This because it is needed to be called externaly, in order to find radio
node in the controller  database.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add handler to channel selection request
Add to the controller, a message handler to channel selection
request from bml.
The message handler sends channel preference query message
to son_slave.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: add bus subscription to new add message
Add subscription to channel preference query message to
backhaul manager.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
Agent: change print format to hex
Changed print format of unknow 1905.1 messages from decimal to hex,
because it is more readable in log files.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

tomereli commented:
No description for the PR - please add one explaining the whole feature, should be a summary of all your changes, why you did them, etc.

tomereli commented:
No line breaks at and of files - please fix that and configure vscode to set it automatically - https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files

tomereli commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -512,6 +512,17 @@ int bml_wfca_controller(BML_CTX ctx, const char *cmd, char *ret_buf, int ret_buf
+ *
+ * @param [in] ctx BML Context.
+ * @param [in] al_mac al_mac of the platform. 
+ * @param [in] ruid Agent radio identifier. 
```
```
@param [in] ruid agent radio identifier
```

morantr replied:
Actually the convention in this file is to capitalize first letter in the description.

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -681,6 +681,15 @@ void cli_bml::setFunctionsMapAndArray()
         static_cast<pFunction>(&cli_bml::bml_wfca_controller_caller), 1, 1, STRING_ARG);
```
Commit message grammar fix:

>Add channel selection command to beerocks_cli_bml command map
and it**s** caller.

morantr commented:
> No description for the PR - please add one explaining the whole feature, should be a summary of all your changes, why you did them, etc.

added description

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1276,6 +1285,14 @@ int cli_bml::bml_wfca_agent_caller(int numOfArgs)
+
+int cli_bml::bml_channel_selection_caller(int numOfArgs)
+{
+    if (numOfArgs == 2) {
```
Maybe we should have the arguments optional - so if one of the arguments is missing, we trigger channel selection for all radios.

morantr replied:
Channel selection is working on one radio at the time. It doesn't make sense to it for all radios.

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1855,4 +1872,4 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
 }
-#endif //BEEROCKS_RDKB
\ No newline at end of file
+#endif //BEEROCKS_RDKB
```
why did this line change?

morantr replied:
This is the last line, and new line was added after this line.

tomereli commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1720,6 +1720,43 @@ void son_management::handle_bml_message(
 #endif //BEEROCKS_RDKB
+
+    case beerocks_message::ACTION_BML_TRIGGER_CHANNEL_SELECTION_REQUEST: {
+
```
remove new line before case statement

tomereli commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -226,6 +226,8 @@ class db {
     std::set<std::string> get_node_siblings(std::string mac, int type = beerocks::TYPE_ANY);
     std::set<std::string> get_node_children(std::string mac, int type = beerocks::TYPE_ANY,
                                             int state = beerocks::STATE_ANY);
+    std::string get_ruid_key_from_parent_mac_and_ruid(std::string parent_mac,
```
this commit is redundant as is this function - just do `parent_mac + "_" + ruid` when you need the key. It is much more clear this way that this is the key. Then add a separate commit which deletes this function completely.

morantr replied:
I think it should be in function. It doesn't seems right to me to concatenate 2 strings in all the places which you have to use key to access a node.
Besides, If one day we want to change the key, we will have to do it in a lot of files.
I think that it is good idea maybe to change the function name to:
 std::string get_node_key(const std::string &al_mac, const std::string &ruid)

tomereli replied:
sounds good - do that then.

tomereli commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1720,6 +1720,43 @@ void son_management::handle_bml_message(
+
+        std::string al_mac   = network_utils::mac_to_string(bml_request->al_mac());
+        std::string ruid_mac = network_utils::mac_to_string(bml_request->ruid());
+        std::string key      = database.get_ruid_key_from_parent_mac_and_ruid(al_mac, ruid_mac);
```
change to
```
std::string key      = al_mac + "_" + ruid_mac;
```

tomereli commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1720,6 +1720,43 @@ void son_management::handle_bml_message(
+        Socket *son_slave_sd = database.get_node_socket(key);
+
+        if (son_slave_sd == nullptr) {
+            LOG(INFO) << "Could not find socket for al_mac=" << al_mac << ", ruid=" << ruid_mac;
```
This is an error. How does the BML know that the operation failed?

morantr replied:
I didn't implement response mechanism. Does response mechanism is required?
I use INFO instead of ERROR because I assumed that the user can send invalid data, and it it just will be ignored.

tomereli replied:
So at least change to ERROR, the user should know that he didn't provide the correct parameters, so at least he will see that in the logs.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4321,4 +4321,4 @@ bool slave_thread::send_cmdu_to_controller(ieee1905_1::CmduMessageTx &cmdu_tx)
                                   backhaul_params.bridge_mac);
-}
\ No newline at end of file
+}
```
Is this just a new line at the end of the file?

morantr replied:
yes

tomereli reviewed (changes requested):
Overall pretty good for your first PR, I have some small comments. Please do fixups and not squash.

morantr committed:
controller: change db function prototype
Changed the name of the function that converts al_mac and ruid to a node
key, and its argument name.
Also, updated all the places it is being called.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
